### ISSUE TITLE
Evaluation bug for top-level closures

### ISSUE THREAD

**Issue by dhil** [links-lang/links#422]
The following program, `bug.links`, triggers an assertion error in the evaluator:
```links
(handle({println("Before `do'"); do Foo(2); println("After `do'") }) { 
  case Return(_) -> fun(x) { x } 
  case Foo(x, resume) -> fun(_) { resume(())(x) } 
})(0)
```
The error:
```
$ ./links --enable-handlers bug.links
Before `do'
*** Error: "Assert_failure core/evalir.ml:80:11"
```
The error happens during a lookup in the `fun_defs` table (`evalir.ml`):
```
let lookup_fun_def f =
    match lookup_fun f with
    | None -> None
    | Some (finfo, _, None, location) ->
      begin
        match location with
        | `Server | `Unknown ->
          (* TODO: perhaps we should actually use env here - and make
             sure we only call this function when it is sufficiently
             small *)
          Some (`FunctionPtr (f, None))
        | `Client ->
          Some (`ClientFunction (Js.var_name_binder (f, finfo)))
        | `Native -> assert false
      end
    | _ -> assert false (* this assertion is triggered *)
```
Curiously, if we thunk the abstractor expression and immediately apply that thunk, then the program runs correctly, i.e.
```links
fun () { 
 handle({println("Before `do'"); do Foo(2); println("After `do'") }) { 
  case Return(_) -> fun(x) { x } 
  case Foo(x, resume) -> fun(_) { resume(())(x) } 
 }
}()(0)
```
Putting the above into a file called `fixed.links` and running it yields the correct result:
```
$ ./links --enable-handlers fixed.links
Before `do'
After `do'
2 : Int
```

**Comment by dhil** [comment]
Here is a simpler example of, what I presume is, the same bug. It does not use handlers:
```links
links> (switch(Foo(id)) { case Foo(id) -> fun(x) { id(x) } })(2);
*** Error: "Assert_failure core/evalir.ml:80:11"
links> switch(Foo(id)) { case Foo(id) -> fun(x) { id(x) } };
*** Error: "Assert_failure core/evalir.ml:80:11"
```
where `id` is the identity function from the prelude. By dropping the tag `Foo`, the program evaluates correctly:
```
links> (switch(id) { case id -> fun(x) { id(x) } })(2);
2 : Int
```

**Comment by frank-emrich** [comment]
I will have a look at it

**Comment by frank-emrich** [comment]
It seems like this has been fixed by 8257f6a1884c20f917f4903fa02b6a08725a7632 and ee3b32a3aa8d67422abe175e8c8b81c4bf774f66.
